<a href="https://colab.research.google.com/github/BrianGisemba/KNN-Naive-Bayes-Classifiers/blob/main/Moringa_School_Week_9_IP_Brian_Michira_Naive_Bayes_Classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Define the Question

##a)Specifying the Question

Predict if an email is a spam or not using the Naive Bayes Classifier.

##b)Defining the Metrics of Success

Creating a model that has an accuracy of above 80%.

##c)Understanding the Context

Spam emails are uninvited bulk-sent messages delivered to an inbox.Email spam has steadily grown since the early 1990's and by 2014 was estimated to account for around 90% of total email traffic.Most email spam messages are commercial in nature.Whether commercial or not,many are not only annoying but also dangerous because they may contain links that lead to punishing websites or sites that are hosting malware or including malware as file attachment.This could lead to cybercrminals gaining access to your computer and other devices.

##d)Recording the Experimental Design

1. Defining the Question.
2. Loading and Previewing the Dataset.
3. Data Cleaning.
4. Modelling.
5. Challenging the Solution.
6. Conclusion.



#Importing Libraries

In [1]:
# importing libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore') 

#Loading and Previewing the Dataset

In [2]:
#loading the dataset
df=pd.read_csv("/content/spambase (1).data")
df.head()

0  0.64  0.64.1  0.1  0.32   0.2  ...   0.43   0.44  3.756   61   278  1
0  0.21  0.28    0.50  0.0  0.14  0.28  ...  0.180  0.048  5.114  101  1028  1
1  0.06  0.00    0.71  0.0  1.23  0.19  ...  0.184  0.010  9.821  485  2259  1
2  0.00  0.00    0.00  0.0  0.63  0.00  ...  0.000  0.000  3.537   40   191  1
3  0.00  0.00    0.00  0.0  0.63  0.00  ...  0.000  0.000  3.537   40   191  1
4  0.00  0.00    0.00  0.0  1.85  0.00  ...  0.000  0.000  3.000   15    54  1

[5 rows x 58 columns]

In [3]:
#listing the variables
columns = ['word_freq_make',
          'word_freq_address',      
          'word_freq_all',          
          'word_freq_3d',          
          'word_freq_our',          
          'word_freq_over',         
          'word_freq_remove',       
          'word_freq_internet',     
          'word_freq_order',        
          'word_freq_mail',         
          'word_freq_receive',      
          'word_freq_will',         
          'word_freq_people',       
          'word_freq_report',       
          'word_freq_addresses',    
          'word_freq_free',         
          'word_freq_business',     
          'word_freq_email',        
          'word_freq_you',          
          'word_freq_credit',       
          'word_freq_your',         
          'word_freq_font',         
          'word_freq_000',          
          'word_freq_money',        
          'word_freq_hp',           
          'word_freq_hpl',          
          'word_freq_george',       
          'word_freq_650',          
          'word_freq_lab',          
          'word_freq_labs',         
          'word_freq_telnet',       
          'word_freq_857',          
          'word_freq_data',         
          'word_freq_415',          
          'word_freq_85',           
          'word_freq_technology',   
          'word_freq_1999',         
          'word_freq_parts',        
          'word_freq_pm',           
          'word_freq_direct',       
          'word_freq_cs',           
          'word_freq_meeting',      
          'word_freq_original',     
          'word_freq_project',      
          'word_freq_re',           
          'word_freq_edu',          
          'word_freq_table',        
          'word_freq_conference',   
          'char_freq_;',            
          'char_freq_(',            
          'char_freq_[',            
          'char_freq_!',            
          'char_freq_$',            
          'char_freq_#',            
          'capital_run_length_average', 
          'capital_run_length_longest', 
          'capital_run_length_total',
          'spam']

In [4]:
#loading the dataset with the column names
df=pd.read_csv("/content/spambase (1).data",names=columns)
df.head()

word_freq_make  word_freq_address  ...  capital_run_length_total  spam
0            0.00               0.64  ...                       278     1
1            0.21               0.28  ...                      1028     1
2            0.06               0.00  ...                      2259     1
3            0.00               0.00  ...                       191     1
4            0.00               0.00  ...                       191     1

[5 rows x 58 columns]

In [5]:
#finding the number of records
df.shape

(4601, 58)

In [6]:
#checking the info of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

#Cleaning the Dataset

## Completness

In [7]:
#checking for missing values
df.isnull().sum()

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

There is no missing values

## Consistency

In [8]:
#Checking for duplicates
df.duplicated().sum()

391

In [9]:
#viewing the duplicates
df[df.duplicated()]

word_freq_make  word_freq_address  ...  capital_run_length_total  spam
26               0.0                0.0  ...                        82     1
103              0.0                0.0  ...                       292     1
104              0.0                0.0  ...                        66     1
105              0.0                0.0  ...                       292     1
106              0.0                0.0  ...                        66     1
...              ...                ...  ...                       ...   ...
4439             0.0                0.0  ...                       282     0
4441             0.0                0.0  ...                       282     0
4537             0.0                0.0  ...                         6     0
4541             0.0                0.0  ...                         2     0
4550             0.0                0.0  ...                         4     0

[391 rows x 58 columns]

In [10]:
#drop duplicated rows
df.drop_duplicates(inplace=True)

In [11]:
#checking if there are any duplicates
df.duplicated().sum()

0

##OUtliers

In [12]:
# dealing with outliers
# quantiles
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR = Q3 - Q1

# Removing outliers based on the IQR range 
df_out= df[~((df < (Q1 - 1.5 * IQR)) | (df > (Q3 + 1.5 * IQR))).any(axis=1)]

# Printing the shape of our new dataset
print(df_out.shape)

# Printing the shape of our old dataset
print(df.shape)

(114, 58)
(4210, 58)


We decided not to drop the outliers since its a large percentage of the whole dataset and it would really affect out analysis.

#Explanatory Data Analysis

In [26]:
#checking the correlation between the variables
independent=df.drop(columns=['spam'],axis=1)
correlation=independent.corr()
correlation

word_freq_make  ...  capital_run_length_total
word_freq_make                    1.000000  ...                  0.084127
word_freq_address                 0.034012  ...                  0.034590
word_freq_all                     0.063207  ...                  0.051961
word_freq_3d                      0.005388  ...                  0.023785
word_freq_our                     0.021740  ...                 -0.010498
word_freq_over                    0.055067  ...                  0.063403
word_freq_remove                  0.011061  ...                 -0.017081
word_freq_internet               -0.004370  ...                  0.036895
word_freq_order                   0.103842  ...                  0.233198
word_freq_mail                    0.040829  ...                  0.078587
word_freq_receive                 0.063031  ...                  0.122057
word_freq_will                    0.075911  ...                  0.004599
word_freq_people                  0.061985  ...                  0.090242
word_freq_report                  0.032263  ...                  0.153939
word_freq_addresses               0.022454  ...                  0.132484
word_freq_free                    0.065443  ...                 -0.003013
word_freq_business                0.081471  ...                  0.057109
word_freq_email                   0.051571  ...                  0.030280
word_freq_you                     0.120563  ...                 -0.027298
word_freq_credit                  0.020924  ...                  0.063523
word_freq_your                    0.121458  ...                  0.047364
word_freq_font                   -0.025574  ...                  0.105218
word_freq_000                     0.126478  ...                  0.157330
word_freq_money                   0.205689  ...                  0.081323
word_freq_hp                     -0.075054  ...                 -0.045403
word_freq_hpl                    -0.063467  ...                 -0.062773
word_freq_george                 -0.048086  ...                 -0.080950
word_freq_650                    -0.050054  ...                 -0.070811
word_freq_lab                    -0.041541  ...                 -0.057905
word_freq_labs                   -0.054241  ...                 -0.067065
word_freq_telnet                 -0.038899  ...                 -0.047253
word_freq_857                    -0.031634  ...                 -0.048248
word_freq_data                   -0.042715  ...                  0.007065
word_freq_415                    -0.026322  ...                 -0.045611
word_freq_85                     -0.046083  ...                 -0.047362
word_freq_technology             -0.057296  ...                 -0.047641
word_freq_1999                   -0.058771  ...                 -0.005995
word_freq_parts                  -0.008626  ...                 -0.015230
word_freq_pm                     -0.011199  ...                 -0.052763
word_freq_direct                 -0.035467  ...                 -0.029950
word_freq_cs                     -0.005280  ...                 -0.026031
word_freq_meeting                -0.026546  ...                 -0.059456
word_freq_original               -0.025603  ...                 -0.044064
word_freq_project                -0.022980  ...                 -0.043103
word_freq_re                     -0.037848  ...                 -0.101349
word_freq_edu                    -0.034455  ...                 -0.049118
word_freq_table                  -0.000950  ...                  0.003558
word_freq_conference             -0.018904  ...                 -0.011919
char_freq_;                      -0.027248  ...                  0.055265
char_freq_(                      -0.019636  ...                  0.112663
char_freq_[                      -0.034784  ...                  0.007228
char_freq_!                       0.059184  ...                  0.025249
char_freq_$                       0.101957  ...                  0.205982
char_freq_#                     

There are no variables that are highly correlated hence we can cotinue with our modelling.

#Modelling

In [14]:
#viewing the columns of the dataset
df.columns

Index(['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d',
       'word_freq_our', 'word_freq_over', 'word_freq_remove',
       'word_freq_internet', 'word_freq_order', 'word_freq_mail',
       'word_freq_receive', 'word_freq_will', 'word_freq_people',
       'word_freq_report', 'word_freq_addresses', 'word_freq_free',
       'word_freq_business', 'word_freq_email', 'word_freq_you',
       'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000',
       'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 'word_freq_george',
       'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet',
       'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85',
       'word_freq_technology', 'word_freq_1999', 'word_freq_parts',
       'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting',
       'word_freq_original', 'word_freq_project', 'word_freq_re',
       'word_freq_edu', 'word_freq_table', 'word_freq_conference',


In [15]:
#Separating the variables
X=df.drop(columns=['spam'],axis=1).values
y=df['spam'].values

In [16]:
X

array([[0.000e+00, 6.400e-01, 6.400e-01, ..., 3.756e+00, 6.100e+01,
        2.780e+02],
       [2.100e-01, 2.800e-01, 5.000e-01, ..., 5.114e+00, 1.010e+02,
        1.028e+03],
       [6.000e-02, 0.000e+00, 7.100e-01, ..., 9.821e+00, 4.850e+02,
        2.259e+03],
       ...,
       [3.000e-01, 0.000e+00, 3.000e-01, ..., 1.404e+00, 6.000e+00,
        1.180e+02],
       [9.600e-01, 0.000e+00, 0.000e+00, ..., 1.147e+00, 5.000e+00,
        7.800e+01],
       [0.000e+00, 0.000e+00, 6.500e-01, ..., 1.250e+00, 5.000e+00,
        4.000e+01]])

In [17]:
y

array([1, 1, 1, ..., 0, 0, 0])

Since our independent variables are continuos we will apply the Gaussian Naive Bayes Classifierto predict whether an email is a Spam or not.

##Gaussian Naive Bayes Classifier

##80:20

In [18]:
# Splitting our data into a training set and a test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0) 

#Feature scaling
from sklearn.preprocessing import StandardScaler  
sc = StandardScaler()  
X_train = sc.fit_transform(X_train)  
X_test = sc.transform(X_test) 

#Training our model
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()  
model = clf.fit(X_train, y_train) 

#Predicting our test predictors
y_pred = model.predict(X_test)

# Evaluating the Algorithm 
from sklearn.metrics import classification_report, confusion_matrix
from sklearn import metrics
print(f'Accuracy score =: {metrics.accuracy_score(y_test, y_pred)} ')
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy score =: 0.8123515439429929 
[[353 142]
 [ 16 331]]
              precision    recall  f1-score   support

           0       0.96      0.71      0.82       495
           1       0.70      0.95      0.81       347

    accuracy                           0.81       842
   macro avg       0.83      0.83      0.81       842
weighted avg       0.85      0.81      0.81       842



Spliting the data into **80:20** leads to an **accuracy score of 81%**.

**True Positive :**353

**True Negative :**331

**False Positive :** 142

**False Negative:**16

##70:30

In [19]:
# Splitting our data into a training set and a test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0) 

#feature scaling
from sklearn.preprocessing import StandardScaler  
sc = StandardScaler()  
X_train = sc.fit_transform(X_train)  
X_test = sc.transform(X_test) 

#Training our model
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()  
model = clf.fit(X_train, y_train) 

#Predicting our test predictors
y_pred = model.predict(X_test)

#Evaluating the Algorithm 
from sklearn.metrics import classification_report, confusion_matrix
from sklearn import metrics
print(f'Accuracy score =: {metrics.accuracy_score(y_test, y_pred)} ')
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy score =: 0.8297703879651623 
[[537 200]
 [ 15 511]]
              precision    recall  f1-score   support

           0       0.97      0.73      0.83       737
           1       0.72      0.97      0.83       526

    accuracy                           0.83      1263
   macro avg       0.85      0.85      0.83      1263
weighted avg       0.87      0.83      0.83      1263



Splitting our data into **70:30**, our model yields an  **accuracy of 83%** which is higher than when we splitted our data into 80:20.

**True Positive :**537

**True Negative :**511

**False Positive :** 200

**False Negative:**15

##60:40

In [20]:
# Splitting our data into a training set and a test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0) 

#feature scaling
from sklearn.preprocessing import StandardScaler  
sc = StandardScaler()  
X_train = sc.fit_transform(X_train)  
X_test = sc.transform(X_test) 

#Training our model
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()  
model = clf.fit(X_train, y_train) 

#Predicting our test predictors
y_pred = model.predict(X_test)

#Evaluating the Algorithm 
from sklearn.metrics import classification_report, confusion_matrix
from sklearn import metrics
print(f'Accuracy score =: {metrics.accuracy_score(y_test, y_pred)} ')
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy score =: 0.8200712589073634 
[[715 279]
 [ 24 666]]
              precision    recall  f1-score   support

           0       0.97      0.72      0.83       994
           1       0.70      0.97      0.81       690

    accuracy                           0.82      1684
   macro avg       0.84      0.84      0.82      1684
weighted avg       0.86      0.82      0.82      1684



Splitting our data into **60:40** our model leads to an **accuracy score of 82%** which is higher than when we spliited the data into 80-20 but less than the accuracy score obtained when splitting the data into 70-30.

**True Positive :**715

**True Negative :**666

**False Positive :** 279

**False Negative:**24

##Hyperparameter Tuning

We decided to carry out hyperparameter tuning to obtain the best parameters to apply in our model.

In [21]:
#carrying out the Gris Search Cross Validation
from sklearn.model_selection import GridSearchCV
nb_classifier = GaussianNB()

params_NB = {'var_smoothing': np.logspace(0,-9, num=100)}
gs_NB = GridSearchCV(estimator=nb_classifier, 
                 param_grid=params_NB, 
                 cv=10,   
                 verbose=1, 
                 scoring='accuracy') 
gs_NB.fit(X_train, y_train)

gs_NB.best_params_

Fitting 10 folds for each of 100 candidates, totalling 1000 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 1000 out of 1000 | elapsed:    3.8s finished


{'var_smoothing': 0.0001873817422860383}

In [22]:
gs_NB.best_estimator_

GaussianNB(priors=None, var_smoothing=0.0001873817422860383)

## Applying the tuned parameters in the 80:20                                                                                    

In [23]:
#Applying the tuned parameters
#Splitting our data into a training set and a test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0) 

#Feature scaling
from sklearn.preprocessing import StandardScaler  
sc = StandardScaler()  
X_train = sc.fit_transform(X_train)  
X_test = sc.transform(X_test) 

#Training our model
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB(var_smoothing=0.0001873817422860383)  
model = clf.fit(X_train, y_train) 

#Predicting our test predictors
y_pred = model.predict(X_test)

#Evaluating the Algorithm 
from sklearn.metrics import classification_report, confusion_matrix
from sklearn import metrics
print(f'Accuracy score =: {metrics.accuracy_score(y_test, y_pred)} ')
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy score =: 0.8123515439429929 
[[353 142]
 [ 16 331]]
              precision    recall  f1-score   support

           0       0.96      0.71      0.82       495
           1       0.70      0.95      0.81       347

    accuracy                           0.81       842
   macro avg       0.83      0.83      0.81       842
weighted avg       0.85      0.81      0.81       842



After applying the tuned parameters we obtained the same accuracy score showing that the tuned parameters had no effect to the accuracy score.

**True Positive :**353

**True Negative :**331

**False Positive :** 142

**False Negative:**16

## Applying the tuned parameters to the 70:30

In [24]:
#Splitting our data into a training set and a test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0) 

#feature scaling
from sklearn.preprocessing import StandardScaler  
sc = StandardScaler()  
X_train = sc.fit_transform(X_train)  
X_test = sc.transform(X_test) 

#Training our model
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB(priors=None, var_smoothing=0.0001873817422860383)  
model = clf.fit(X_train, y_train) 

#Predicting our test predictors
y_pred = model.predict(X_test)

# Evaluating the Algorithm 
from sklearn.metrics import classification_report, confusion_matrix
from sklearn import metrics
print(f'Accuracy score =: {metrics.accuracy_score(y_test, y_pred)} ')
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy score =: 0.8297703879651623 
[[537 200]
 [ 15 511]]
              precision    recall  f1-score   support

           0       0.97      0.73      0.83       737
           1       0.72      0.97      0.83       526

    accuracy                           0.83      1263
   macro avg       0.85      0.85      0.83      1263
weighted avg       0.87      0.83      0.83      1263



After applying the tuned parameters we obtained the same accuracy score showing that the tuned parameters had no effect to the accuracy score.

**True Positive :**537

**True Negative :**511

**False Positive :** 200

**False Negative:**15

## Applying the tuned parameters to the 60:40

In [25]:
#Splitting our data into a training set and a test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0) 

#feature scaling
from sklearn.preprocessing import StandardScaler  
sc = StandardScaler()  
X_train = sc.fit_transform(X_train)  
X_test = sc.transform(X_test) 

#Training our model
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB(priors=None, var_smoothing=0.0001873817422860383)  
model = clf.fit(X_train, y_train) 

#Predicting our test predictors
y_pred = model.predict(X_test)

#Evaluating the Algorithm 
from sklearn.metrics import classification_report, confusion_matrix
from sklearn import metrics
print(f'Accuracy score =: {metrics.accuracy_score(y_test, y_pred)} ')
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy score =: 0.8254156769596199 
[[725 269]
 [ 25 665]]
              precision    recall  f1-score   support

           0       0.97      0.73      0.83       994
           1       0.71      0.96      0.82       690

    accuracy                           0.83      1684
   macro avg       0.84      0.85      0.83      1684
weighted avg       0.86      0.83      0.83      1684



Applying the tuned parameters leads to an increase in the accuracy score **82.0% to 82.5%**.

**True Positive :**725

**True Negative :**665

**False Positive :** 269

**False Negative:**25

#Challenging the Solution

1. Did we have the right data?Yes.The data provided was right.
2. Did we use the correct models?Yes.The model had an accuracy score above 80%.

## Conclusion

**Model Accuracy**

**Before Tuning the parameters**

**80:20 -**81%

**70:30 -**83%

**60:40 -**82%

**After Tuning the parameters**

**80:20 -**81%

**70:30 -**82%

**60:40 -**82.5%


Splitting the data into 70:30 yields to a higher accuracy score of 83%.Hence we recommended that the model should be used in predicting if the message is a spam or not.